tests/test_models.py

```diff
@@ -59,11 +60,22 @@ class QcdCatalogDemographicModelTestMixin(CatalogDemographicModelTestMixin):
     Extends the tests to also check that the qc model is equal to
     the production model.
     """
-    # To be defined in subclass.
-    qc_model = None
 
     def test_qc_model_equal(self):
-        self.assertTrue(self.model.equals(self.qc_model))
+        self.assertTrue(self.model.equals(self.model.qc_model))
+
+
+# Add model specific test classes, derived from one of the above.
+for species in stdpopsim.all_species():
+    for model in species.demographic_models:
+        superclasses = [unittest.TestCase]
+        if model.qc_model is not None:
+            superclasses.append(QcdCatalogDemographicModelTestMixin)
+        else:
+            superclasses.append(CatalogDemographicModelTestMixin)
+        classname = f"Test{species.id}{model.id}"
```

Comment:
Nice! Generating classes at run-time like a pro.

Reply:
Could do some basic sanity checks here, like this
```python
qc_test_classes = []
for species in stdpopsim.all_species():
￼    for model in species.demographic_models:
￼        superclasses = [unittest.TestCase]
￼        if model.qc_model is not None:
￼            superclasses.append(QcdCatalogDemographicModelTestMixin)
￼        else:
￼            superclasses.append(CatalogDemographicModelTestMixin)
￼        classname = f"Test{species.id}{model.id}"
        cls = type(classname, tuple(superclasses), dict(model=model))

# Do some basic sanity checks here to double check no errors are introduced 
# leading to tests not being run silently
assert len(qc_test_classes) > 0
for cls in qc_test_classes:
    assert issubclass(cls, unittest.TestCase)
    setattr(sys.modules[__name__], cls.__name__, cls)
```

Reply:
I agree this could catch some kinds of bugs, so I've added it in. I don't think we can check that tests are being run in a more general case though, because it's not clear that `nose` and/or `unittest` wont change how they find tests. In particular, they could decide to register classes upon creation using a metaclass, and I'm not sure that dynamic creation using `type` would catch that (but then, I know almost nothing about metaclasses).

To give an example that it's not necessarily obvious how tests are identified... I first tried programmatically adding `test_{species.id}_{model.id}` methods to a `TestQCModels(unittest.TestCase)` class. I did this using `setattr(TestQCModels, ...)` after the class definition. Tests defined manually within the TestQCModels class ran as normal, but none of those added with setattr ran. I briefly looked at the code for unittest.TestCase, and didn't see anything obviously tricky there, so gave up and did the dynamic-class version instead (which is probably nicer anyway).

Reply:
Using metaclasses is nearly always a bad idea...